subprojects/core/src/main/java/org/gradle/cache/internal/ProducerGuard.java

```diff
@@ -15,8 +15,127 @@
+ * This is <b>not</b> a cache. The factory should take care of caching wherever it makes sense.
+ *
+ * <p>
+ * The concurrency level and memory usage depend on the implementation, see {@link #mapBased()}, {@link #striped()} and {@link #serial()} for details.
+ * </p>
+ *
+ * @param <T> the type of key to lock on
+ */
+public abstract class ProducerGuard<T> {
+
+    /**
+     * Creates a {@link ProducerGuard} which guarantees that different keys will block on different locks,
+     * ensuring maximum concurrency. This comes at the cost of allocating locks for each key, leading to
+     * relatively high memory pressure. If the above guarantee is not necessary, consider using a {@link #striped()}
+     * guard instead.
+     */
+    public static <T> ProducerGuard<T> mapBased() {
+        return new MapBasedProducerGuard<T>();
+    }
+
+    /**
+     * Creates a {@link ProducerGuard} which evenly spreads calls over a fixed number of locks.
+     * This means that in some cases two different keys can block on the same lock. The benefit of
+     * this strategy is that it uses only a fixed amount of memory. If your code depends on
+     * different keys always getting different locks, use a {@link #mapBased()} guard instead.
+     */
+    public static <T> ProducerGuard<T> striped() {
+        return new StripedProducerGuard<T>();
+    }
+
+    /**
+     * Creates a {@link ProducerGuard} which limits concurrency to a single factory at a time,
+     * ignoring the key. This is mainly useful for testing.
+     */
+    public static <T> ProducerGuard<T> serial() {
+        return new SerialProducerGuard<T>();
+    }
+
+    private ProducerGuard() {
+
+    }
+
+    /**
+     * Runs the given factory, guarded by the given key.
+     *
+     * @param key the key to lock on
+     * @param factory the factory to run under the lock
+     * @param <V> the type returned by the factory
+     * @return the value returned by the factory
+     */
+    public final <V> V guardByKey(T key, Factory<V> factory) {
+        Lock lock = getLock(key);
+        try {
+            lock.lock();
+            return factory.create();
+        } finally {
+            lock.unlock();
+            returnLock(key);
+        }
+    }
+
+    protected abstract Lock getLock(T key);
+
+    protected abstract void returnLock(T key);
+
+    private static class MapBasedProducerGuard<T> extends ProducerGuard<T> {
+        private final ConcurrentMap<T, Lock> locks = new ConcurrentHashMap<T, Lock>();
+
+        @Override
+        protected Lock getLock(T key) {
+            Lock lock = locks.get(key);
+            if (lock == null) {
+                ReentrantLock newLock = new ReentrantLock();
+                lock = locks.putIfAbsent(key, newLock);
+                if (lock == null) {
+                    lock = newLock;
+                }
+            }
+            return lock;
+        }
+
+        @Override
+        protected void returnLock(T key) {
+            locks.remove(key);
```

Comment:
Good catch! I really should have written the unit test before changing the existing implementation. There is now a test demonstrating that the contract is honored.

I could make the map-based approach work by using a cache with weak keys and never calling `remove()` explictily. But that would lead to more memory pressure, so I instead reverted back to your algorithm. 

The main performance gain of this PR was introducing the striped implementation for file system snapshotting (which is called much more often and suffered from the synchronization overhead). If the synchronization ever becomes a problem in dependency resoultion too, we can reevaluate the weak cache.